File: semestr_3/aisd/z02_z03/W1Rekurencja/silnia.cpp
#include <iostream>
using namespace std;

int rec_factorial(int n){
    if (n <= 0)
        return 1;
    return n * rec_factorial(n-1);
}

int iter_factorial(int n){
    int result = 1;
    for (int i = 1; i < n+1; i++)
        result = result * i;
    return result;
}

int tail_factorial(int n, int accumulator = 1){
    if(n <= 0)
        return accumulator;
    return tail_factorial(n-1, accumulator * n);
}

int main(){
    int a = rec_factorial(10);
    int b = rec_factorial(0);
    int c = rec_factorial(-2);
    //Rekursja
    cout<<a<<" "<<b<<" "<<c<<endl;

    a = iter_factorial(10);
    b = iter_factorial(0);
    c = iter_factorial(-2);
    //Iteracja
    cout<<a<<" "<<b<<" "<<c<<endl;

    a = tail_factorial(10,-2);
    b = tail_factorial(0, 3);
    c = tail_factorial(-2,5);
    //Iteracja
    cout<<a<<" "<<b<<" "<<c<<endl;
}

File: semestr_3/aisd/z02_z03/W1Rekurencja/silnia.py
def rec_factorial(n):
    if(n <= 0):
        return 1
    return n * rec_factorial(n-1)

def iter_factorial(n):
    result = 1
    for i in range(1,n+1):
        result = result * i
    return result

def tail_factorial(n, accumulator = 1):
    if(n <= 0):
        return accumulator
    return tail_factorial(n-1, accumulator * n)

a = rec_factorial(10)
b = rec_factorial(0)
c = rec_factorial(-2)
print(a, b, c)

a = iter_factorial(10)
b = iter_factorial(0)
c = iter_factorial(-2)
print(a, b, c)

# a =rec_factorial(-1)
# import sys
# print(sys.getrecursionlimit()) #1000
# # Blokada by nie dopuścić by program wykonywał się w nieskończoność
